[MHV-61146] sets focus on error field after clicking send btn

vsaleem opened the pull request:
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Are you removing, renaming or moving a folder in this PR?
- [ ] No, I'm not changing any folders (skip to Summary and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

If the folder you changed contains a `manifest.json`, search for its `entryName` in the content-build [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) (the `entryName` there will match).

If an entry for this folder exists in content-build and you are:
1. **Deleting a folder**:
   1. First search `vets-website` for _all_ instances of the `entryName` in your `manifest.json` and remove them in a separate PR. Look particularly for references in `src/applications/static-pages/static-pages-entry.js` and `src/platform/forms/constants.js`. _**If you do not do this, other applications will break!**_
      - _Add the link to your merged vets-website PR here_
   2. Then, Delete the application entry in [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) and merge that PR **before** this one
      - _Add the link to your merged content-build PR here_

2. **Renaming or moving a folder**: Update the entry in the [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json), but do not merge it until your vets-website changes here are merged. The content-build PR must be merged immediately after your vets-website change is merged in to avoid CI errors with content-build (and Tugboat).

### :warning: TeamSites :warning:
Examples of a TeamSite: https://va.gov/health and https://benefits.va.gov/benefits/. This scenario is also referred to as the "injected" header and footer. You can reach out in the `#sitewide-public-websites` Slack channel for questions.

Did you change site-wide styles, platform utilities or other infrastructure?
- [ ] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary

- This fix resolves the focus issue on compose form and electronic signature validation error fields
- After clicking 'Send' or 'Save draft' buttons on blank fields, the focus should revert the first validation error/errors on screen.
- If user continues to click on the 'Send' or 'Save draft' button when there are validation errors on screen, then the focus to revert to that error.

### User Story:  
As a SM user, I want to see the focus land on the correct error after clicking 'Send' or 'Save draft on a blank draft, so that I can complete all fields on the compose form before sending or saving a message.

Given: Given I am composing a secure message, and I have made no changes/updates to the draft 
When: When I attempt to 'Send' or 'Save draft' without completing any fields
Then:  Then I should see the focus revert to the first validation error available and hear the announcement on the screenreader.

### Acceptance Criteria:
AC1   Validation errors should appear on all compose fields and on electronic signature box (if TG that requires a signature is selected)
AC2  Focus should revert to first available validation error on screen until non are present.
AC3:  UCD validation

## Related issue(s)

-  [[MHV-61146]](https://jira.devops.va.gov/browse/MHV-61146) Resolve focus on validation errors on compose fields & electronic signature fields

## Testing done

- [ ] Automation Testing
- [ ] Manual testing
- [ ] 508 Validation

## Screenshots

_Note: This field is mandatory for UI changes (non-component work should NOT have screenshots)._

|         | Before | After |
| ------- | ------ | ----- |
| Mobile  |        |       |
| Desktop |        |       |

## What areas of the site does it impact?

*Va.gov - Secure Messaging*

## Acceptance criteria

### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

Commit:
[MHV-61146] sets focus on error field after clicking send btn

fazilqa committed:
uncomment failed tests

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

oleksii-morgun commented on src/applications/mhv-secure-messaging/tests/e2e/secure-messaging-compose-errors.cypress.spec.js:
```diff
@@ -19,8 +18,8 @@ describe('Secure Messaging Compose Errors', () => {
-    // PatientComposePage.verifyErrorText(Data.PLEASE_SELECT_RECIPIENT);
-    // PatientComposePage.verifyFocusOnErrorMessage();
+    PatientComposePage.verifyErrorText(Data.PLEASE_SELECT_RECIPIENT);
+    PatientComposePage.verifyFocusOnErrorMessage();
```
@fazilqa I noticed this method validates that focus is on an error field which may be any one of these on the page. How can we validate that focus is on the right error field? I think that may be a more proper way of validating here. I suggest either rewriting this `verifyFocusOnErrorMessage` function where you can probably pass an element as an argument and validate that focus is present or just simply validate focus in the test without a helper function.  Same applies everywhere else in this PR where this function is used

fazilqa replied:
done

alex-morgun-1 reviewed (changes requested):
@vsaleem looks good 👍🏻 I do have some suggestions for cypress coverage, @fazilqa please see my comment below

fazilqa committed:
update verifyFocusOnErrorMessage

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
[MHV-61146] Resolve recipient select bug, user is not allowed to send if recipient is not selected

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
[MHV-61146] Bug fix, prevents user from saving or sending a message  to specific tg if all required fields aren't completed

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

fazilqa committed:
add test to compose-errors spec

fazilqa committed:
Merge remote-tracking branch 'origin/61146-Resolve-focus-on-validation-errs-sm' into 61146-Resolve-focus-on-validation-errs-sm

fazilqa committed:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
[MHV-61146] Accessibility fix to adjust focus to announce the recipient field and electronic signature required alert correctly

Commit:
[MHV-61146] Unit test fixes in progress

fazilqa committed:
test fix

Commit:
[MHV-61146] unit test updates for electronic signature checkbox

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -264,14 +273,34 @@ const ComposeForm = props => {
+          focusOnErrorField();
+        }
+      };
+      handleEffect();
```
did you run in a particular scenario where you would not benefit from calling the code in `handleEffect` const directly without wrapping in a const?

vsaleem replied:
Actually, yes! I would get a strange `DOMException` error that failed to execute the `'querySelector'` within the `focusElement()` helper.  I chose to refactor this code this way so we can use `async await` anyway so that the SR focus would announce the selected `recipient` first before announcing the `signature-alert`.

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -264,14 +273,34 @@ const ComposeForm = props => {
+          selectedTG
+        ) {
+          await focusElement(selectedTG?.name);
+          focusElement(alertRef.current);
```
interesting, what are you trying to achieve with calling one `focusElement` right after another?

vsaleem replied:
https://github.com/department-of-veterans-affairs/vets-website/pull/31543#discussion_r1744542971

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -351,16 +380,14 @@ const ComposeForm = props => {
       if (
-        (isSignatureRequired && !electronicSignature) ||
-        isSignatureRequired === null
+        messageValid === true &&
```
```suggestion
        messageValid &&
```

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -351,16 +380,14 @@ const ComposeForm = props => {
-        (isSignatureRequired && !checkboxMarked) ||
-        isSignatureRequired === null
-      ) {
+      if (messageValid === true && isSignatureRequired && !checkboxMarked) {
```
```suggestion
      if (messageValid && isSignatureRequired && !checkboxMarked) {
```

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -499,14 +525,15 @@ const ComposeForm = props => {
       await setSendMessageFlag(false);
       const validSignatureNotRequired = messageValid && !isSignatureRequired;
       const validSignatureRequired =
-        isSignatureRequired && signatureValid && checkboxValid;
+        isSignatureRequired && signatureValid && checkboxValid && messageValid;
```
I am wondering if renaming `validSignatureRequired` to `validSignature` or `isSignatureValid` ay be more comprehensive? I am just finding it a bit confusing when contrasted with `validSignatureNotRequired`.

KyleCardwell commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -264,14 +273,34 @@ const ComposeForm = props => {
+          selectedTG
+        ) {
+          await focusElement(selectedTG?.name);
+          focusElement(alertRef.current);
```
Can you help me understand lines 286 and 287?  What is the purpose of focusing on an element then immediately focusing on something else?

vsaleem replied:
@KyleCardwell This was done so that screen readers announced the selected recipient first before announcing the alert. Normally, when an alert is displayed it gets read immediately. However in this case, we still need to guarantee that the user is aware of their recipient selection before announcing the alert.

oleksii-morgun replied:
wouldn't `aria-describedby` be a better implementation of this sequence? @BobbyBaileyRB

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/ComposeForm.jsx:
```diff
@@ -517,19 +544,19 @@ const ComposeForm = props => {
         messageBody === '' &&
         subject === '' &&
-        (selectedRecipient === 0 || selectedRecipient === '0') &&
+        (selectedRecipientId === 0 || selectedRecipientId === '0') &&
```
```suggestion
        Number(selectedRecipientId) === 0 &&
```

oleksii-morgun commented on src/applications/mhv-secure-messaging/components/ComposeForm/RecipientsSelect.jsx:
```diff
@@ -78,13 +79,26 @@ const RecipientsSelect = ({
+      const recipient = recipientsList?.find(r => +r.id === +e.detail.value);
+
+      if (e.detail.value === '') {
+        setSelectedRecipient({});
```
Just curious why do we need to set this to a blank object `{}` if normal values are strings? should it be `null` or `' '` ?

vsaleem replied:
Having `null` or an `' '` here caused a small bug that would not focus on the select field when the value was deselected. We need the blank `{}` here because `selectedRecipient` and `recipient` are returning objects when selected. If `e.detail.value` has an empty value, `' '`, then the `selectedRecipient` should be an empty `{}`.

oleksii-morgun replied:
understood. As as we tested during code pairing, we did ensure that selectedRecipient is populated with an object, so that makes sense

oleksii-morgun commented on src/applications/mhv-secure-messaging/tests/components/Compose/ComposeForm.unit.spec.jsx:
```diff
@@ -891,9 +892,13 @@ describe('Compose form component', () => {
+      <ComposeForm
+        recipients={initialState.sm.recipients}
+        messageValid
+        isSignatureRequired
```
I dont think `messageValid` and `isSignatureRequired` are valid props of this component?

oleksii-morgun commented on src/applications/mhv-secure-messaging/tests/components/Compose/ComposeForm.unit.spec.jsx:
```diff
@@ -929,17 +934,66 @@ describe('Compose form component', () => {
+      <ComposeForm
+        recipients={initialState.sm.recipients}
+        messageValid
+        isSignatureRequired
```
same as above

oleksii-morgun commented on src/applications/mhv-secure-messaging/tests/components/Compose/RecipientsSelect.unit.spec.jsx:
```diff
@@ -57,10 +57,12 @@ describe('RecipientsSelect', () => {
       onValueChange,
       setCheckboxMarked,
       setElectronicSignature,
+      setAlertDisplayed,
```
Did you mean to add validation for this spy?

Commit:
cleanup

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
cleanup

fazilqa committed:
fix failed tests

fazilqa committed:
code cleanup

fazilqa committed:
update test

Commit:
cleanup

Commit:
[MHV-61146] cleanup; Removed separate signature and checkbox validation checks

- all validation check will now be operated by messageValid

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
[MHV-61146] Resolve bug; focus remains on recipient after selection and user can confirm saving draft without electronic signature

- all compose form and electronic signature fields must be valid before user can save draft

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
[MHV-61146] Remove aria-live polite from alert to prevent announcing close button

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

fazilqa committed:
update method according to correct behavior

Commit:
[MHV-61146] Bug fix, user can save draft with attachment modal

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

vsaleem commented on src/applications/mhv-secure-messaging/tests/components/Compose/ComposeForm.unit.spec.jsx:
```diff
@@ -920,17 +921,62 @@ describe('Compose form component', () => {
+        messageBody,
+      );
+    });
 
```
Electronic signature and checkbox errors no longer require all other fields to be populated before displaying.. Removing all of this text

Commit:
[MHV-61146] Cleanup unit tests and ElectronicSignature fields must be populated before sending a msg fix

Commit:
[MHV-61146] cleanup

Commit:
cleanup

fazilqa committed:
update focus assertion

fazilqa committed:
update constants

fazilqa committed:
Merge remote-tracking branch 'origin/61146-Resolve-focus-on-validation-errs-sm' into 61146-Resolve-focus-on-validation-errs-sm

# Conflicts:

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
checking for interpolation errors

Commit:
[MHV-61146] Adding cy test updates

Commit:
Merge branch 'main' into 61146-Resolve-validation-focus-3rd-branch

Commit:
Merge branch 'main' into 61146-Resolve-validation-focus-3rd-branch

Commit:
[MHV-61146] Fix validation focus on Reply page

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch '61146-Resolve-validation-focus-3rd-branch' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Update secure-messaging-keyboard-nav-save-draft.cypress.spec.js

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm

Commit:
Merge branch 'main' into 61146-Resolve-focus-on-validation-errs-sm
